### ISSUE TITLE
`span` throwing error on 32-bit system with greater than 2 gig span

### ISSUE THREAD

**Issue by OvermindDL1** [Microsoft/GSL#696]
Getting random crashes/halts while running a program built to run on some embedded hardware (no OS, it runs straight) with 4 gigs of ram, 32-bit, tracking it down to a span constructor when the span was being fed an allocated array that was greater than 2 gigs in size, tracked it down to span trying to narrow it to a range that is half the size of the range of allocateable memory (the full 32-bit range of the 4 gigs of RAM is available here) then consequently halting the embedded chip here.

In short, weeded it down to `span` storing as a signed value, which makes no sense...  How is this class supposed to be used if it can't even represent the representeable memory range of the hardware this is compiled on?  What is the reasoning for breaking it in such a way?  Where are the usecases in being able to store negative spans?  How is this class supposed to be used on such embedded hardware (especially as it is is being passed to another library that I do not control)?

**Comment by OvermindDL1** [comment]
Got me thinking, shouldn't I be seeing an unsigned to sign cast warning at compile-time?  But I'm not, I wonder if this has something to do with it...

https://github.com/Microsoft/GSL/blob/master/include/gsl/span#L68

Still though, even if I got a warning (though it would have saved me a lot of troubleshooting time before I found the broken signed-issue...) I need to know how span can be used for greater than half the addressable pointer space.

**Comment by OvermindDL1** [comment]
I can see that `std::ptrdiff_t` is 'supposed' to be larger than the largest representable range for the compiled platform, and this toolchain I have may have slight issues with the newer standards, so I'll go ahead and close this.

As an aside though, being a signed type is still wrong.  If a sentinel value is wanted to represent dynamically sized spans then the maximum value for the representable pointer (which could be double sized) for the given pointer size should be used (which is -1 as well by happenchance, but just in an unsigned context), along with disallowing signed usage in the constructors and methods via overloads (or runtime checks on those), that would be the safest method as it would help to dissuade improper size calculations unlike the current spec that seems to encourage improper calculations.
